sdk/inc/azure/iot/az_iot_hub_client.h

```diff
@@ -75,7 +75,12 @@ AZ_NODISCARD az_iot_hub_client_options az_iot_hub_client_options_default();
  *
  * @param[out] client The #az_iot_hub_client to use for this call.
  * @param[in] iot_hub_hostname The IoT Hub Hostname.
- * @param[in] device_id The Device ID.
+ * @param[in] device_id The Device ID. If the ID contains any of the following characters, they must

```

Comment:
What is the **data format** of the device id? Is it base64 encoded? Is it a Guid? Is it a number?

Reply:
I'm not sure if we need to go that far into detail. I think the minute specifics of the documentation are delegated to the service side docs which detail how IoT works in general. The "device id" is a veryyy fundamental and well known construct in IoT. If we are to go to that level of documentation here, tbh it's going to required a bunchhh of work and would be a headache.

Reply:
Aside from whether we want to document that or not, for my edification, can you explain what the data format is?

Reply:
Sure it's a name given to the device by the user, typically containing mostly letters but could be numbers as well. Standard "C string" encoded as UTF-8 here.

Reply:
How long can it be? Can it contain any non-ASCII characters? What about other special characters like `#%/:=?[\]^`{|}`

Reply:
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-identity-registry

A case-sensitive string (up to 128 characters long) of ASCII 7-bit alphanumeric characters plus certain special characters: - . + % _ # * ? ! ( ) , : = @ $ '.

Reply:
Is there a link/resource where I can read up on which of these characters need to be percent encoding?

It kind-of looks like the device ID needs to be URL encoded.